lib/using_restricted_array.rb

```diff
@@ -1,61 +1,130 @@
-require_relative 'restricted_array.rb'
+require_relative "restricted_array.rb"
+require "pry"
 # RestrictedArray can be created using a specified size, or a random size in
 # the range of 1-20 will be chosen for you.
 # All values are integers in the range of 1-221.
 # RestrictedArray cannot be resized.
 
+# NOTES----
+# TIME COMPLEXITY: Linear or O(n), where n is the number of items in the array or the length of the input array. The big O indicates we're considering the worst case scenario of how long the algorithm will take to complete the task.
+# IMPORTANT NOTE:
+# Constant O(1); (ignore, time it takes to run is fixed)
+# Linear (add complexity together)
+# (n*20) when 20 is a constant set in the loop --> if there's a constant, drop the constant and say that the time complexity is O(n).)
+# Quadratic (multiply when loops are nested inside of one another; O(n*n) or O(n^3))
+
+# SPACE COMPLEXITY:
+# Contant: O(1)
+# Linear: O(n) <-- amount of space taken to loop through something takes the same amount of space (i takes 2 bytes to hold the counter, max (variable) takes 2 bites to hold the current max value, etc.)
+# Quadratic: O(n*m) duel btwn jedis (m) and siths (n), a winner is determined
+# -----
+
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(1) constant because i always takes up the same amount of space; the method is not dependant upon the length of the
+
 def length(array)
-  raise NotImplementedError
+  i = 0
+  until array[i] == nil
+    # also valid --> while array[i] != nil
+    i += 1
+  end
+  return i
 end
 
-# Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Prints each integer value in the array
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(1)
 def print_array(array)
-  raise NotImplementedError
+  length = length(array)
+  i = 0
+  while i < length
+    print array[i]
+    print " "
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of the array
+# Space complexity: O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    return true if array[i] == value_to_find
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of the array
+# Space complexity: O(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  largest_value = array[0]
+  i = 0
+  length.times do |item|
+    (largest_value = array[i]) if (array[i] > largest_value)
```

Comment:
Nice! 😄

FYI: you can also do `largest_value = max(largest_value, array[i])` if you _really_ want to be concise.